enhancements/network/multi-network-vrfs.md

```diff
@@ -0,0 +1,924 @@
+single router VRF aware, which adds more complexity into OVN.
+
+The proposal here is to create a GR per network. With this topology, OVN will create a patch port per network to the
+br-ex bridge. OVN-Kubernetes will be responsible for being VRF/network aware and forwarding packets via flows in br-ex
+to the right GR. Each per-network GR will only have load balancers configured on it for its network, and only be able to
+route to pods in its network. The logical topology would look something like this, if we use an example of having a
+cluster default primary network, a layer 3 primary network, and a layer 2 primary network:
+
+![VRF Topology](images/VRFs.svg)
+
+In the above diagram, each network is assigned a unique conntrack zone and conntrack mark. These are required in order
+to be able to handle overlapping networks egressing into the same VRF and SNAT’ing to the host IP. Note, the default
+cluster network does not need to use a unique CT mark or zone, and will continue to work as it does today. This is due
+to the fact that no user-defined network may overlap with the default cluster subnet. More details in the next section.
+
+#### Shared Gateway Mode
+
+##### Pod Egress
+
+On pod egress, the respective GR of that network will handle doing the SNAT to a unique masquerade subnet IP assigned to
+this network. For example, in the above diagram packets leaving GR-layer3 would be SNAT’ed to 169.254.169.5 in zone 64005.
+The packet will then enter br-ex, where flows in br-ex will match this packet, and then SNAT the packet to the node IP
+in zone 0, and apply its CT mark of 5. Finally, the packet will be recirculated back to table 0, where the packet will
+be CT marked with 1 in zone 64000, and sent out of the physical interface. In OVN-Kubernetes we use zone 64000 to track
+things from OVN or the host and additionally, we mark packets from OVN with a CT Mark of 1 and packets from the host
+with 2. Pseudo openflow rules would look like this (assuming node IP of 172.18.0.3):
+
+```
+pod-->GR(snat, 169.254.169.5, zone 64005)->br-ex(snat, 172.18.0.3, zone 0, mark 5, table=0) -->recirc table0 (commit
+zone 64000, mark 1) -->eth0
+```
+
+The above design will accommodate for overlapping networks with overlapping ports. The worst case scenario is if two
+networks share the same address space, and two pods with identical IPs are trying to connect externally using the same
+source and destination port. Although unlikely, we have to plan for this type of scenario. When each pod tries to send a
+packet through their respective GR, SNAT’ing to the unique GR masquerade IP differentiates the conntrack entries. Now,
+when the final SNAT occurs in br-ex with zone 0, they can be determined as different connections via source IP, and when
+SNAT’ing to host IP, conntrack will detect a collision using the same layer 4 port, and choose a different port to use.
```

Comment:
collision can still happen today, which if you remember from a long time ago we had with shared gw mode with the host and OVN not using the same conntrack zone. The new part that this new design is covering is when pods have overlapping IP addresses. The collision referenced in this doc is slightly different. So the old type of collision would be this:

pod A (10.244.0.5:55000) ---->GR (SNAT 172.18.0.4)---> 8.8.8.8:53
pod B (10.244.0.6:55000) ---->GR (SNAT 172.18.0.4)---> 8.8.8.8:53

In the above case both connections look the same as they leave the node. 172.18.0.4:55000 -> 8.8.8.8:53. This is a collision because the server (8.8.8.8) will see 2 different TCP sessions talking as if its in the same TCP session. So the kernel will recognize this and modify the 2nd connection's source port, so now you have, for example:
172.18.0.4:55000 -> 8.8.8.8:53
172.18.0.4:55001 -> 8.8.8.8:53

With this enhancement, now you can have overlapping IP addresses:
pod A (10.244.0.6:55000) ---->GR (SNAT 172.18.0.4)---> 8.8.8.8:53
pod B (10.244.0.6:55000) ---->GR (SNAT 172.18.0.4)---> 8.8.8.8:53

So when Linux goes to SNAT this in the same zone, it sees these as the same exact session (collision). So we need Linux in zone 0 to realize these are two different conntrack sessions. So to do that we can add an extra SNAT:
pod A (10.244.0.6:55000) ---->ovn_cluster_router (SNAT 169.254.169.5) ---->GR (SNAT 172.18.0.4)---> 8.8.8.8:53
pod B (10.244.0.6:55000) ---->ovn_cluster_router (SNAT 169.254.169.6) ---->GR (SNAT 172.18.0.4)---> 8.8.8.8:53

This is basically a consequence of us route leaking into the default VRF and SNAT'ing to the host IP in that VRF.

Reply:
> collision can still happen today, which if you remember from a long time ago we had with shared gw mode with the host and OVN not using the same conntrack zone. 

oh right! then we did zone0 on gr

> The new part that this new design is covering is when pods have overlapping IP addresses. The collision referenced in this doc is slightly different. So the old type of collision would be this:
> 
> pod A (10.244.0.5:55000) ---->GR (SNAT 172.18.0.4)---> 8.8.8.8:53 pod B (10.244.0.6:55000) ---->GR (SNAT 172.18.0.4)---> 8.8.8.8:53

yup 

> 
> In the above case both connections look the same as they leave the node. 172.18.0.4:55000 -> 8.8.8.8:53. This is a collision because the server (8.8.8.8) will see 2 different TCP sessions talking as if its in the same TCP session. So the kernel will recognize this and modify the 2nd connection's source port, so now you have, for example: 172.18.0.4:55000 -> 8.8.8.8:53 172.18.0.4:55001 -> 8.8.8.8:53

yup yup makes sense and exactly what I was thinking.

> 
> With this enhancement, now you can have overlapping IP addresses: pod A (10.244.0.6:55000) ---->GR (SNAT 172.18.0.4)---> 8.8.8.8:53 pod B (10.244.0.6:55000) ---->GR (SNAT 172.18.0.4)---> 8.8.8.8:53
> 
> So when Linux goes to SNAT this in the same zone, it sees these as the same exact session (collision). So we need Linux in zone 0 to realize these are two different conntrack sessions. So to do that we can add an extra SNAT: pod A (10.244.0.6:55000) ---->ovn_cluster_router (SNAT 169.254.169.5) ---->GR (SNAT 172.18.0.4)---> 8.8.8.8:53 pod B (10.244.0.6:55000) ---->ovn_cluster_router (SNAT 169.254.169.6) ---->GR (SNAT 172.18.0.4)---> 8.8.8.8:53
> 
> This is basically a consequence of us route leaking into the default VRF and SNAT'ing to the host IP in that VRF.

AH gotcha so "conntrack will detect a collision using the same layer 4 port," that is the part which confused me, its not just same layer4 its also gonna have layer3 conflict but conntrack at the end of the day detects it for "same tuple" as a conflict so I guess even if 2 different pods with sameIPs sent traffic, if the ports were different it would be considered ok, yeah I get it. thank you so much Tim for explaining this piece to me